### PR TITLE
Newmetric: ClassificationReport

### LINKED ISSUE #2580: Classification Report, maybe with the option to select even more metrics
## 🚀 Feature

Classification Report, maybe with the option to select even more metrics
### Motivation

Like in sklearn,  the feature of an classification report.

https://scikit-learn.org/stable/modules/generated/sklearn.metrics.classification_report.html

### Pitch

Implementation of an classification report like in Sklearn.

### PR DESCRIPTION
**Pull request by aymuos15**
## What does this PR do?

Fixes #2580

<details>
  <summary>Before submitting</summary>

- [x] Was this **discussed/agreed** via a Github issue? (no need for typos and docs improvements) - Yes, details in issue number mentioned.
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/torchmetrics/blob/master/.github/CONTRIBUTING.md), Pull Request section? - Yes.
- [ ] Did you make sure to **update the docs**? - Not yet. Can do once everything is finalised in terms of code.
- [x] Did you write any new **necessary tests**? - Yes

</details>

<details>
  <summary>PR review</summary>

Classification report is a nice to have metric. The motivation can be easily inspired from its scikit learn integration. Having it ready for tensors is definitely a good step forward.

Notes: 
1. The original issue talks about having the option to include more metrics. I personally feel that is not required as this should be a quick stand alone way to get a lot of relevant (and usually required metrics info) quickly in a nice presentable way. 
2. Maybe we need a more in depth discussion to expand the testing. I do not have access to a distributed mission at the moment, but if it is paramount, I can look into it. 

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

</details>

## Did you have fun?

Was a really nice way of getting familiar with the codebase :)

----
📚 Documentation preview 📚: https://torchmetrics--3116.org.readthedocs.build/en/3116/

### CODE REVIEW

**Review by rittik9 — CHANGES_REQUESTED** [review]
@aymuos15  I had a quick look and opened a pr for fixing pre-commit errors.
Doctests regarding classification report seem to be failing, mind having look.
going to take a detailed look later.

File: src/torchmetrics/functional/classification/classification_report.py [hunk 1]
```diff
@@ -0,0 +1,638 @@
...
+        >>> target = tensor([0, 1, 0, 1, 0, 1])
+        >>> preds = tensor([1, 0, 1, 1, 0, 1])
+        >>> metric = ClassificationReport(
+        ...     task="binary",
+        ...     num_classes=2,
+        ...     output_dict=False,
+        ... )
+        >>> metric.update(preds, target)
+        >>> test_result = metric.compute()
+        >>> print(test_result)
```
Comment (rittik9) [inline src/torchmetrics/functional/classification/classification_report.py]: try using `pprint` instead of `print` for doctests.

Comment (aymuos15) [reply src/torchmetrics/functional/classification/classification_report.py]: Thanks a lot for the suggestion. This actually helped me single out the actual error.

However, using pprint forces one to kinda use a lot of quotes and newlines characters. The normal print works like scikit learn as well.

Is this okay or should I just change everything to pprint?

**Review by rittik9 — COMMENTED** [review]
Hi @aymuos15 pls update the  _init_.py files in the functional interface and docs.
You can refer https://github.com/Lightning-AI/torchmetrics/pull/3090

**Comment by aymuos15** [comment]
Thank for letting me know. Will get that done.

Just had a question, should I make changes to `docs/source/links.rst` and `CHANGELOG.md` as well?

**Comment by rittik9** [comment]
updating `docs/source/links.rst` not necessary in this case. But pls update the `CHANGELOG.md`

**Comment by aymuos15** [comment]
I having a bit of difficulty developing the docs locally, hence the multiple commits for the doc fixes.

While im aware of: https://github.com/Lightning-AI/pytorch-lightning/issues/16158

I get:

```bash
ERROR: Could not find a version that satisfies the requirement lai-sphinx-theme (from versions: none)

[notice] A new release of pip is available: 24.3.1 -> 25.1.1
[notice] To update, run: pip install --upgrade pip
ERROR: No matching distribution found for lai-sphinx-theme
```

When I run 
```bash
pip install -e . --quiet -r requirements/_docs.txt
```

Is there a better way to go about this? Thanks!

**Comment by rittik9** [comment]
Hi @Borda @aymuos15 , I'll try to review it by tomorrow...

File: src/torchmetrics/functional/classification/classification_report.py [hunk 2]
```diff
@@ -0,0 +1,644 @@
...
+
+def _compute_averages(
+    class_metrics: Dict[str, Dict[str, Union[float, int]]],
+) -> Dict[str, Dict[str, Union[float, int]]]:
+    """Compute macro and weighted averages for the classification report."""
+    total_support = int(sum(metrics["support"] for metrics in class_metrics.values()))
+    num_classes = len(class_metrics)
+
+    averages: Dict[str, Dict[str, Union[float, int]]] = {}
+    for avg_name in ["macro avg", "weighted avg"]:
```
Comment (rittik9) [inline src/torchmetrics/functional/classification/classification_report.py]: let's add 'micro' as well

File: src/torchmetrics/functional/classification/classification_report.py [hunk 3]
```diff
@@ -0,0 +1,644 @@
...
+        <BLANKLINE>
+            accuracy                           0.78         6
+           macro avg       0.83      0.83      0.83         6
+        weighted avg       0.83      0.83      0.83         6
+
+    """
+    # Compute task-specific metrics
+    if task == ClassificationTask.BINARY:
+        class_metrics = _compute_binary_metrics(preds, target, threshold, validate_args)
+        accuracy_val = binary_accuracy(preds, target, threshold, validate_args=validate_args).item()
```
Comment (rittik9) [inline src/torchmetrics/functional/classification/classification_report.py]: for accuracy as well can we show class wise and micro macro and weighted? let's repeat it for multiclass and multilabel as well.

File: src/torchmetrics/functional/classification/classification_report.py [hunk 4]
```diff
@@ -0,0 +1,644 @@
...
+def _compute_binary_metrics(
+    preds: Tensor, target: Tensor, threshold: float, validate_args: bool
+) -> Dict[int, Dict[str, Union[float, int]]]:
+    """Compute metrics for binary classification."""
+    class_metrics = {}
+
+    for class_idx in [0, 1]:
+        if class_idx == 0:
+            # Invert for class 0 (negative class)
+            inv_preds = 1 - preds
```
Comment (rittik9) [inline src/torchmetrics/functional/classification/classification_report.py]: in stead of these can we use the 'ignore_index' argument?

File: src/torchmetrics/functional/classification/classification_report.py [hunk 5]
```diff
@@ -0,0 +1,644 @@
...
+        threshold=threshold,
+        target_names=target_names,
+        digits=digits,
+        output_dict=output_dict,
+        zero_division=zero_division,
+        validate_args=validate_args,
+    )
+
+
+def multiclass_classification_report(
```
Comment (rittik9) [inline src/torchmetrics/functional/classification/classification_report.py]: can we also provide top_k for multiclass?

**Review by rittik9 — COMMENTED** [review]
please remove the unnecessary comments

File: src/torchmetrics/classification/classification_report.py
Comment (rittik9) [inline src/torchmetrics/classification/classification_report.py:1]: In case of modular interface , while computing  , why not call the functions from functional interface, reducing boilerplate

**Comment by aymuos15** [comment]
@rittik9 Thanks again for the detailed review. 

I have tried to address everything bar the classwise accuracy score. That is inherently different from the scikit version of the metric. Are you okay if we raise a separate issue for that, discuss how exactly we want to add that, and then do a separate PR for that?

**Comment by aymuos15** [comment]
Thank you very much for the review. 

I definitely like the idea and agree with it. 

However, I think I would have to pretty much revamp everything. Would it be okay to keep this as the base and proceed with that separately in another PR or should I just continue here?

**Comment by Borda** [comment]
> However, I think I would have to pretty much revamp everything. Would it be okay to keep this as the base and proceed with that separately in another PR or should I just continue here?

you can continue here 🐰

**Comment by rittik9** [comment]
Hi @aymuos15 While you're at it, could you please ensure that the 'micro' averaging option is included by default for all three cases—binary, multiclass, and multilabel? In case the results don't align with sklearn, we can always verify them manually.

**Comment by Borda** [comment]
> Since anyways now we are going to include everything, ill make all options exhaustive.

We can figure out how to split it into smaller pieces so it would land smoother/fater

**Comment by aymuos15** [comment]
That sounds good with me. Could you please let me know what that would entail?

This is what i had in mind:

```bash
TorchMetrics Classification Report Examples

============================================================
 BINARY CLASSIFICATION
============================================================
Predictions: [0, 1, 1, 0, 1, 0, 1, 1]
True Labels: [0, 1, 0, 0, 1, 1, 1, 0]

--- CASE 1: DEFAULT METRICS (precision, recall, f1-score) ---
             precision    recall  f1-score   support 

           0      0.67      0.50      0.57         4 
           1      0.60      0.75      0.67         4 

    accuracy                          0.62         8 
   micro avg      0.62      0.62      0.62         8 
   macro avg      0.63      0.62      0.62         8 
weighted avg      0.63      0.62      0.62         8 

--- CASE 2: ADD ACCURACY TO DEFAULT METRICS ---
             precision    recall  f1-score  accuracy   support 

           0      0.67      0.50      0.57      0.50         4 
           1      0.60      0.75      0.67      0.75         4 

    accuracy                                    0.62         8 
   micro avg      0.62      0.62      0.62      0.62         8 
   macro avg      0.63      0.62      0.62      0.62         8 
weighted avg      0.63      0.62      0.62      0.62         8 

--- CASE 3: SPECIFICITY ONLY ---
             specificity     support 

           0        0.75           4 
           1        0.50           4 

    accuracy        0.62           8 
   micro avg        0.50           8 
   macro avg        0.62           8 
weighted avg        0.62           8 

Verification - Direct accuracy calculation: 0.6250

============================================================
 MULTICLASS CLASSIFICATION
============================================================
Predictions: [0, 2, 1, 2, 0, 1, 2, 0]
True Labels: [0, 1, 1, 2, 0, 2, 2, 1]

--- CASE 1: DEFAULT METRICS ---
             precision    recall  f1-score   support 

           0      0.67      1.00      0.80         2 
           1      0.50      0.33      0.40         3 
           2      0.67      0.67      0.67         3 

    accuracy                          0.62         8 
   micro avg      0.62      0.62      0.62         8 
   macro avg      0.61      0.67      0.62         8 
weighted avg      0.60      0.63      0.60         8 

--- CASE 2: ADD ACCURACY TO DEFAULT METRICS ---
             precision    recall  f1-score  accuracy   support 

           0      0.67      1.00      0.80      1.00         2 
           1      0.50      0.33      0.40      0.33         3 
           2      0.67      0.67      0.67      0.67         3 

    accuracy                                    0.62         8 
   micro avg      0.62      0.62      0.62      0.62         8 
   macro avg      0.61      0.67      0.62      0.67         8 
weighted avg      0.60      0.63      0.60      0.63         8 

--- CASE 3: SPECIFICITY ONLY ---
             specificity     support 

           0        0.83           2 
           1        0.80           3 
           2        0.80           3 

    accuracy        0.62           8 
   micro avg        0.81           8 
   macro avg        0.81           8 
weighted avg        0.81           8 

Verification - Direct accuracy calculation: 0.6667

============================================================
 MULTILABEL CLASSIFICATION
============================================================
Predictions:
  Sample 1: [1, 0, 1]
  Sample 2: [0, 1, 0]
  Sample 3: [1, 1, 0]
  Sample 4: [0, 0, 1]
True Labels:
  Sample 1: [1, 0, 0]
  Sample 2: [0, 1, 1]
  Sample 3: [1, 1, 0]
  Sample 4: [0, 0, 1]

--- CASE 1: DEFAULT METRICS ---
             precision    recall  f1-score   support 

           0      1.00      1.00      1.00         2 
           1      1.00      1.00      1.00         2 
           2      0.50      0.50      0.50         2 

   micro avg      0.83      0.83      0.83         6 
   macro avg      0.83      0.83      0.83         6 
weighted avg      0.83      0.83      0.83         6 
 samples avg      0.88      0.88      0.83         6 

--- CASE 2: ADD ACCURACY TO DEFAULT METRICS ---
             precision    recall  f1-score  accuracy   support 

           0      1.00      1.00      1.00      1.00         2 
           1      1.00      1.00      1.00      1.00         2 
           2      0.50      0.50      0.50      0.50         2 

   micro avg      0.83      0.83      0.83      0.83         6 
   macro avg      0.83      0.83      0.83      0.83         6 
weighted avg      0.83      0.83      0.83      0.83         6 
 samples avg      0.88      0.88      0.83      0.83         6 

--- CASE 3: SPECIFICITY ONLY ---
             specificity     support 

           0        1.00           2 
           1        1.00           2 
           2        0.50           2 

   micro avg        0.83           6 
   macro avg        0.83           6 
weighted avg        0.83           6 
 samples avg                       6 

Verification - Direct accuracy calculation: 0.8333
```

Essentially
1. If anyone uses it off the shelf, it is exactly like scikit-learn
2. But then, we have the option to add whatever has been discussed till now as well.

Is the above okay?

I have not committed anything yet because the code is very messy. Once we agree on a path forward, I will trigger the next commit if thats okay.

**Comment by aymuos15** [comment]
Ah okay! I will push a commit for the formatting and printing tonight. Thank you.

**Comment by Borda** [comment]
How is it going here?

**Comment by aymuos15** [comment]
Hi @Borda Didn't realise this was in the pipeline still. Happy to revisit this. Do you think itll be okay if I gave it a go with the collection style to complete the whole PR? Thanks.
